File: Darts/src/code/screen/game/MatchScorer.java
package code.screen.game;

import bean.AbstractTableRenderer;
import code.db.DartsMatchEntity;
import code.db.ParticipantEntity;
import code.utils.DartsColour;

/**
 * For the 'Match Summary' tab.
 */
public final class MatchScorer extends AbstractScorer
{
	private static final int COLUMN_NO_GAME_ID = 0;
	private static final int COLUMN_NO_FINAL_SCORE = 1;
	private static final int COLUMN_NO_FINISHING_POSITION = 2;
	private static final int COLUMN_NO_MATCH_POINTS = 3;
	
	private DartsMatchEntity match = null;
	
	/**
	 * Game #, Score, Position, Points
	 */
	@Override public int getNumberOfColumns()
	{
		return 4;
	}
	
	@Override
	public void initImpl(String gameParams)
	{
		tableScores.setGameColumnIndex(0);
		
		for (int i=COLUMN_NO_GAME_ID + 1; i<model.getColumnCount(); i++)
		{
			tableScores.getColumn(i).setCellRenderer(new ParticipantRenderer(i));
		}
		
		tableScores.setColumnWidths("100");
	}
	
	public void setMatch(DartsMatchEntity match)
	{
		this.match = match;
	}
	
	public void updateResult()
	{
		int totalScore = 0;
		
		int rowCount = tableScores.getRowCount();
		for (int i=0; i<rowCount; i++)
		{
			ParticipantEntity pt = (ParticipantEntity)tableScores.getValueAt(i, COLUMN_NO_MATCH_POINTS);
			totalScore += match.getScoreForFinishingPosition(pt.getFinishingPosition());
		}
		
		lblResult.setVisible(true);
		lblResult.setText("" + totalScore);
		
		//Also update the screen
		tableScores.repaint();
		
		if (match.isComplete())
		{
			//updateResultColourForPosition(rowCount);
		}
	}
	
	/**
	 * Inner classes
	 */
	private final class ParticipantRenderer extends AbstractTableRenderer<ParticipantEntity>
	{
		private int colNo = -1;
		
		public ParticipantRenderer(int colNo)
		{
			this.colNo = colNo;
		}
		
		@Override
		public Object getReplacementValue(ParticipantEntity pt)
		{
			if (colNo == COLUMN_NO_FINAL_SCORE)
			{
				int score = pt.getFinalScore();
				if (score == -1)
				{
					return "N/A";
				}
				
				return score;
			}
			else if (colNo == COLUMN_NO_FINISHING_POSITION)
			{
				return pt.getFinishingPositionDesc();
			}
			else if (colNo == COLUMN_NO_MATCH_POINTS)
			{
				return match.getScoreForFinishingPosition(pt.getFinishingPosition());
			}
			else
			{
				return "";
			}
		}
		
		@Override
		public void setCellColours(ParticipantEntity typedValue, boolean isSelected)
		{
			if (colNo == COLUMN_NO_FINISHING_POSITION)
			{
				int finishingPos = typedValue.getFinishingPosition();
				DartsColour.setFgAndBgColoursForPosition(this, finishingPos);
		
			}
		}
	}
}

File: Darts/src/main/kotlin/dartzee/screen/game/scorer/MatchScorer.kt
package dartzee.screen.game.scorer

import dartzee.db.DartsMatchEntity
import dartzee.db.ParticipantEntity
import dartzee.utils.DartsColour
import dartzee.core.bean.AbstractTableRenderer

/**
 * For the 'Match Summary' tab.
 */
class MatchScorer : AbstractScorer()
{
    private var match: DartsMatchEntity? = null

    /**
     * Game #, Score, Position, Points
     */
    override fun getNumberOfColumns() = 4

    override fun initImpl(gameParams: String)
    {
        tableScores.setLinkColumnIndex(0)

        for (i in COLUMN_NO_GAME_ID + 1 until model.columnCount)
        {
            tableScores.getColumn(i).cellRenderer = ParticipantRenderer(i)
        }

        tableScores.setColumnWidths("100")
    }

    fun setMatch(match: DartsMatchEntity)
    {
        this.match = match
    }

    fun updateResult()
    {
        var totalScore = 0

        val rowCount = tableScores.rowCount
        for (i in 0 until rowCount)
        {
            val pt = tableScores.getValueAt(i, COLUMN_NO_MATCH_POINTS) as ParticipantEntity
            totalScore += match!!.getScoreForFinishingPosition(pt.finishingPosition)
        }

        lblResult.isVisible = true
        lblResult.text = "" + totalScore

        //Also update the screen
        tableScores.repaint()
    }

    /**
     * Inner classes
     */
    private inner class ParticipantRenderer(private val colNo: Int) : AbstractTableRenderer<ParticipantEntity>()
    {
        override fun getReplacementValue(value: ParticipantEntity): Any
        {
            return when (colNo)
            {
                COLUMN_NO_FINAL_SCORE -> if (value.finalScore == -1) "N/A" else value.finalScore
                COLUMN_NO_FINISHING_POSITION -> value.getFinishingPositionDesc()
                COLUMN_NO_MATCH_POINTS -> match!!.getScoreForFinishingPosition(value.finishingPosition)
                else -> ""
            }
        }

        override fun setCellColours(typedValue: ParticipantEntity?, isSelected: Boolean)
        {
            if (colNo == COLUMN_NO_FINISHING_POSITION)
            {
                val finishingPos = typedValue!!.finishingPosition
                DartsColour.setFgAndBgColoursForPosition(this, finishingPos)
            }
        }
    }

    companion object
    {
        private const val COLUMN_NO_GAME_ID = 0
        private const val COLUMN_NO_FINAL_SCORE = 1
        private const val COLUMN_NO_FINISHING_POSITION = 2
        private const val COLUMN_NO_MATCH_POINTS = 3
    }
}
